magento/pwa-studio
# Handling `https-proxy-agent` GitHub security issue.

## Description

This fixes the [`https-proxy-agent` security issue](https://github.com/magento/pwa-studio/network/alert/yarn.lock/https-proxy-agent/open).

`https-proxy-agent` is a deep dependency of 3 of our dependencies:
1. Danger
2. Graphql CLI
3. Lerna

A GitHub security issue has been raised on `https-proxy-agent@2.2.3 and lower versions`.

Lerna is being used in the `monorepo-introduction.js` script. @zetlen made changes to the files so we don't need lerna any more. Hence deprecating it.

Also `danger` uses `https-proxy-agent@2.2.1` which is vulnerable as well. It will be moved out of `pwa-studio` into the CICD repo where it belongs in (PWA-543)[https://jira.corp.magento.com/browse/PWA-543].

Coming to `graphql-cli` I have tried upgrading it to the latest version but, the deep dependency of `https-proxy-agent` is still `2.2.1`. Not sure how to fix that.

## Related Issue
Closes [PWA-531](https://jira.corp.magento.com/browse/PWA-531)

### Verification Stakeholders
@zetlen 

### Verification Steps
The root `package.json` has a `"prepare"` script defined which runs `scripts/monorepo-introduction.js`. NPM runs the `prepare` script after every local `yarn install`. The purpose of `scripts/monorepo-introduction.js` is to set up the development environment correctly for brand new checkouts, and to verify that all packages are prepared to build. To do so, it runs the `prepare` script in all of the sub-packages that have one defined. It was using Lerna to do this, but now it doesn't need to.

To verify that `scripts/monorepo-introduction.js` is still working correctly:
1. Check out the branch and run `yarn install`.

2. Observe that after the install completes, it logs:
    ```
    Preparing packages...
    Ensuring valid environment...
    ```
    This is the expected output of `scripts/monorepo-introduction.js`.

3. Run `rm packages/graphql-cli-validate-magento-pwa-queries/lib/magento-compatibility.js`.
4. Run `mv packages/venia-concept/.env packages/venia-concept/.env.temp`
5. Run `yarn install` again. Observe that both of the removed files have been regenerated.
6. Run `mv packages/venia-concept/.env.temp packages/venia-concept/.env` to restore your old `.env` file.

To verify that `scripts/monorepo-introduction.js` will run any newly defined `prepare` step in a sub-package:

1. Edit `packages/venia-concept/package.json`.
2. Add `"prepare": "[[ 1 == 2 ]]",` to the `scripts` section. (That's designed to fail on purpose.)
3. Run `yarn install`. Observe that it fails and displays an error.
4. Remove the `prepare` step from `packages/venia-concept/package.json`.

## Checklist
TODO
